storage/backup-restore/src/lib.rs

```diff
@@ -1,7 +1,85 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-#[allow(dead_code)]
-mod adapter;
+pub mod adapter;
 
-type FileHandle = String;
+use crate::adapter::Adapter;
+use anyhow::Result;
+use futures::{stream, StreamExt};
+use libra_crypto::HashValue;
+use libra_types::{proof::SparseMerkleRangeProof, transaction::Version};
+use storage_client::{StorageRead, StorageReadServiceClient};
+
+pub type FileHandle = String;
+
+pub async fn backup_account_state(
+    client: &StorageReadServiceClient,
+    version: Version,
+    adapter: &impl Adapter,
+    max_chunk_size: usize,
+) -> Result<Vec<(FileHandle, FileHandle)>> {
+    let mut chunk = vec![];
+    let mut ret = vec![];
+
+    let mut account_stream = client.backup_account_state_async(version)?;
+    let mut prev_key = None;
+    while let Some(resp) = account_stream.next().await.transpose()? {
+        let key = resp.account_key;
+        let blob = resp.account_state_blob;
+        println!("Backing up key: {:x}", key);
+
+        let mut bytes = key.to_vec();
+        let blob: Vec<u8> = blob.into();
+        assert!(blob.len() <= std::u32::MAX as usize);
+        let blob_len = blob.len() as u32;
+        bytes.extend_from_slice(&blob_len.to_le_bytes());
+        bytes.extend(blob);
+
+        if chunk.len() + bytes.len() > max_chunk_size {
+            assert!(chunk.len() <= max_chunk_size);
+            let account_state_file = adapter
+                .write_new_file(stream::once(async move { chunk }))
+                .await?;
+
+            let prev_key = prev_key.expect("max_chunk_size should be larger than account size.");
+            println!(
+                "Reached max_chunk_size. Asking proof for key: {:?}",
+                prev_key,
+            );
+            let proof_file = get_proof_and_write(client, adapter, prev_key, version).await?;
+            ret.push((account_state_file, proof_file));
+            chunk = vec![];
+        }
+
+        chunk.extend(bytes);
+        prev_key = Some(key);
+    }
+
+    assert!(!chunk.is_empty());
+    assert!(chunk.len() <= max_chunk_size);
+    let account_state_file = adapter
+        .write_new_file(stream::once(async move { chunk }))
+        .await?;
+
+    let prev_key = prev_key.expect("Should have at least one account.");
+    println!("Asking proof for last key: {:x}", prev_key);
```

Comment:
why do we have proof only for the last key?  if we want to give someway for user to verify the backup,  we can just store the root hash for the entire tree?

Also , for the importer to use this backup database (and verify its integrity) , looks like it needs to know the validator set as well, is there an easy way to embed validator set here too?

Reply:
It's not just the last key. We have proof for every chunk of accounts. So when you restore, it's like restore 1000 accounts, check proof1, restore the next 1000 accounts, check proof2, and so on. This way you verify each chunk before writing them to DB.

Reply:
validator set is outside the scope of this change.